File: ag-ui/integrations/claude-agent-sdk/python/examples/agents/tool_based_generative_ui.py
"""
Tool-based generative UI agent configuration.

This module demonstrates how frontend tools (defined by the client) are dynamically
added to Claude and can be called to render UI components.

The key feature: tools are provided by the CLIENT via RunAgentInput.tools,
and Claude can discover and call them without backend implementation.
"""

from ag_ui_claude_sdk import ClaudeAgentAdapter
from .constants import DEFAULT_DISALLOWED_TOOLS


def create_tool_based_generative_ui_adapter() -> ClaudeAgentAdapter:
    """Create adapter for tool-based generative UI demo."""
    system_prompt = """You are a creative writing assistant that renders content using beautiful UI components.

## CRITICAL: Always Use Frontend Tools

When the user asks for creative content (haikus, poems, stories), you MUST use the 
available frontend tools to render them. DO NOT just write the content as text.

### Workflow for Haiku Requests

When the user asks for a haiku, you MUST:
1. Create the haiku (Japanese and English versions)
2. **IMMEDIATELY call the `generate_haiku` tool** with:
   - japanese: array of 3 lines in Japanese (or English if you don't know Japanese)
   - english: array of 3 lines in English  
   - image_name: Pick ONE from the available images (cherry blossoms, Mt Fuji, temples, etc)
   - gradient: CSS gradient for background (e.g., "linear-gradient(135deg, #667eea 0%, #764ba2 100%)")
3. After the tool returns, respond briefly: "I've created a beautiful haiku for you! 🎋"

### IMPORTANT Rules

- **ALWAYS call the tool FIRST** - don't write the haiku as plain text
- The tool will handle the beautiful rendering
- After calling the tool, just give a brief confirmation
- If the user asks for non-creative content, respond normally (no tool needed)

### Example Flow

User: "Write me a haiku about nature"
You: [Call generate_haiku tool with the haiku data]
You: "I've created a beautiful haiku about nature for you! 🎋"

User: "What's 2+2?"
You: "That's 4!" (no tool needed)
"""
    
    return ClaudeAgentAdapter(
        name="tool_based_generative_ui",
        description="Creative writing assistant with frontend tool rendering",
        options={
            "model": "claude-haiku-4-5",
            "system_prompt": system_prompt,
            "disallowed_tools": list(DEFAULT_DISALLOWED_TOOLS),
        }
    )

File: ag-ui/integrations/aws-strands/typescript/examples/server/api/tool-based-generative-ui.ts
/**
 * Tool-based Generative UI example for AWS Strands (TypeScript).
 *
 * The `generate_haiku` tool is declared on the frontend via `useFrontendTool`
 * — the @ag-ui/aws-strands adapter auto-registers it as a proxy tool when
 * `RunAgentInput.tools` arrives, so the backend does not register a native
 * tool here. Strands invokes the proxy with the structured haiku args, the
 * adapter halts the run after the proxy returns, and the browser renders the
 * haiku card from the streamed `TOOL_CALL_*` events.
 */

import { Agent } from "@strands-agents/sdk";
import { StrandsAgent } from "@ag-ui/aws-strands";
import { createStrandsApp } from "@ag-ui/aws-strands/server";
import { createModel } from "../model-factory";

async function main(): Promise<void> {
  const strandsAgent = new Agent({
    model: await createModel(),
    tools: [],
    systemPrompt: `You are a creative haiku generator.

When the user asks for a haiku, ALWAYS call the \`generate_haiku\` tool with:
- 3 lines of haiku in Japanese
- 3 lines of haiku translated to English
- One relevant image_name from the provided list
- A CSS gradient for the card background

Do not respond with plain text — always use the tool.`,
  });

  const aguiAgent = new StrandsAgent({
    agent: strandsAgent,
    name: "tool_based_generative_ui",
    description: "AWS Strands haiku generator with frontend-rendered tool",
  });

  const app = await createStrandsApp(aguiAgent, { path: "/" });
  const port = Number(process.env.PORT ?? 8000);
  app.listen(port, () => {
    console.log(`Listening on http://localhost:${port}`);
  });
}

void main();
